Implement Lean 4 support

insightmind opened the pull request:
# General

This pull request adds support for Lean 4 using [LeanInk](https://github.com/insightmind/LeanInk).

In order to do so, it introduces a new driver `lean4.py` for handling communication between Alectryon and LeanInk.
LeanInk takes a `.lean` file as its input and analyzes it using direct access to the Lean 4 compilers AST. It returns an output file that contains all fragments (Text and Sentences) in Alectryon's json format.

# Usage

## Lean code files
Lean 4 usage is equivalent to any other supported language. You can use the `lean4` or `lean4+rst` driver to explicitly specify Lean 4.
With this PR `.lean` files will also automatically be inferred as Lean 4 files instead of Lean 3.

```
alectryon --frontend lean4 plain-lean4.lean
```

To print the debug output of LeanInk you can do the following:

```
alectryon --frontend lean4 plain-lean4.lean --debug
```

If you want to use Alectryon with a Lean 4 file that uses the `Lake` package manager you can provide the `lakefile` to Alectryon:

```
alectryon --frontend lean4 plain-lean4.lean --lake lakefile.lean
```

## Literate files

Lean 4 is also supported in literate files using the `lean4::` directive for reStructuredText and `{lean4}` for MyST markdown files.

Finally we also added a short explainer section about Lean 4 support in the readme.

# New Recipes

We have added a few new recipes to test and showcase the Lean 4 support:
- `plain-lean4.lean`
- `literate-lean4.lean`
- `lean4-tactics.rst`
- `lean4-tactics-myst.md`

These files make up most of the added files.

# Important notes

To support Lean 4 we had to make minor changes to the implementation of Lean 3 support.
We renamed the recipe `plain.lean` to `plain-lean3.lean`.
Additionally `.lean` files will automatically be inferred as Lean 4 files instead of Lean 3 because it has a higher precedence.

# Feedback

Feedback on the implementation is always welcome. Especially if you can spot rough issues or find features that are not yet supported.

Most line additions are due to new recipe files. If you take a look at the python files, the actual implementation is very small.

cpitclaudel commented on alectryon/docutils.py:
```diff
@@ -798,6 +802,10 @@ class Lean3Directive(ProverDirective):
     name = "lean3"
 
+class Lean4Directive(ProverDirective):
+    """Highlight and annotate a Coq snippet."""
```
s/Coq/Lean4/ in that docstring (and ideally /Lean3/ in the one above ^^)

cpitclaudel commented on alectryon/lean4.py:
```diff
@@ -0,0 +1,84 @@
+# Copyright © 2019 Clément Pit-Claudel
```
Change the copyright line to your own name :)

cpitclaudel commented on alectryon/literate.py:
```diff
@@ -862,15 +862,32 @@ def rst2lean3(rst):
+    lit_open=r"/-|", lit_close=r"|-/",
+    lit_open_re=r"[/][-][|][ \t]*", lit_close_re=r"[ \t]*[|]?[-][/]\Z",
+    quote_pairs=[("/-", r"/\ -"), ("-/", r"-\ /")]
+)
```
Oh my, I had forgotten that we had support for literate comments in Lean3.  Does it actually work?  I wrote that code in a giant rush a few months back.

insightmind replied:
It seems to work fine. At least I didn't have any problems with it during my testings of the Lean 4 implementation.

cpitclaudel commented:
Excellent work @insightmind!  I will make a detailed review this week-end, most likely on Sunday (please ping me Saturday evening if you haven't heard back by then).  I just had a quick read-through and everything looks very reasonable, so I don't think there will be much to change before merging :)

cpitclaudel commented:
Btw, since Lean support line comments, we could consider using line comments for literate programming in Lean.  That's how I did it in F* a while back, see https://github.com/FStarLang/fstar-mode.el/tree/master/etc/fslit/ .  If you think that would be a good fit for Lean then we could import the state machine from fslit.

Oh, and feel free to suggest something better than `/-|`, I chose the `|` in Coq because it was relatively easy to type with `(*` and because `(**`, `(*!`, and `(*+` were already taken.

Kha commented on recipes/literate_lean4.lean:
```diff
@@ -0,0 +1,63 @@
+Alectryon supports literate programs and documents (combinations of code and prose) written in Lean4 and reStructuredText.  Here is an example written in Lean4.  It can be converted to reST, HTML, or LaTeX using the following commands::
+
+   alectryon literate_lean4.lean
+       # Coq+reST → HTML;  produces ‘literate_lean4.html’
```
```suggestion
       # Lean4+reST → HTML;  produces ‘literate_lean4.html’
```
etc., I assume

Kha commented:
> Oh, and feel free to suggest something better than `/-|`, I chose the `|` in Coq because it was relatively easy to type with `(*` and because `(**`, `(*!`, and `(*+` were already taken.

I don't mind `/-|`, though for consistency with docstrings `/-- ... -/` and module docstrings `/-! ... -/`, it would be nice to use plain `-/` as the terminator.

insightmind commented:
> Btw, since Lean support line comments, we could consider using line comments for literate programming in Lean. That's how I did it in F* a while back, see https://github.com/FStarLang/fstar-mode.el/tree/master/etc/fslit/ . If you think that would be a good fit for Lean then we could import the state machine from fslit.
> 
> Oh, and feel free to suggest something better than `/-|`, I chose the `|` in Coq because it was relatively easy to type with `(*` and because `(**`, `(*!`, and `(*+` were already taken.

We discussed the idea recently and came to the conclusion that multiline comments are a better fit for Lean. The reason we think so is that Lean already supports docstrings and module docstrings, as described above, using multiline comments. Literate programming using multiline comments is therefore more consistent with the current language.

Kha commented:
> please ping me Saturday evening if you haven't heard back by then

Slightly later ping :)

cpitclaudel commented on alectryon/cli.py:
```diff
@@ -378,6 +378,7 @@ def write_file(ext, strip):
 EXTENSIONS_BY_LANGUAGE = {
     "coq": (".v",),
+    "lean4": (".lean",),
     "lean3": (".lean", ".lean3"),
```
With `.lean` appearing under both `lean3` and `lean4` lists the `CODE_EXTENSIONS` list below will contain `.lean` twice, which may cause issues.  A quick review suggests that `CODE_EXTENSIONS` could simply be turned into a set.  We should review other uses of `EXTENSIONS_BY_LANGUAGE`, too, like in the definition of `FRONTENDS_BY_EXTENSION`.  It may be enough to remove the lean3 entry from that explicitly after constructing the list.

insightmind replied:
Would removing the `.lean` extension from the `.lean3` case be enough?

cpitclaudel replied:
Yes, let's do that.

insightmind replied:
Ok, adjusted it.

cpitclaudel commented on README.rst:
```diff
@@ -523,6 +523,36 @@ Support for quoting snippets and displaying or hiding sentences is partial.
+------
+
+Alectryon also has support for the upcoming Lean 4. To use Alectryon with your Lean 4 files, you have to install LeanInk first. Take a look at `LeanInk <https://github.com/insightmind/LeanInk>`_ for installation instructions.
+Lean 4 support is similar to Lean 3 support. The reStructuredText directive for Lean 4 is ``.. lean4::``, for Markdown/MyST files it is ``{lean4}``. The literate delimiter is also ``/-|``:
```
Let's move the `lean4` section first.  `lean3` support is not complete at all.

cpitclaudel replied:
Also, the colon at the end of this sentence doesn't work with the next sentence

insightmind replied:
Fixed.

cpitclaudel commented on alectryon/cli.py:
```diff
@@ -637,11 +638,18 @@ def post_process_arguments(parser, args):
 
+    leanInk_args = []
+    if args.leanInk_args_lake != None:
+        leanInk_args.extend(("--lake", args.leanInk_args_lake))
```
Will things fail gracefully if this file doesn't exist?

insightmind replied:
Yes and no, if the file does not exist it will not automatically fail, instead it returns to the default behavior of having no lake file, meaning LeanInk will only load the standard library of Lean.

cpitclaudel commented on alectryon/docutils.py:
```diff
@@ -32,6 +32,10 @@
+    .. lean4::
+
+        #check Nat
+
```
Let's remove these three examples; that will be one less spot to update when we add other languages

insightmind replied:
Removed.

cpitclaudel commented on alectryon/docutils.py:
```diff
@@ -48,7 +52,7 @@
 ``alectryon-container``).
 
-Inline Coq highlighting is provided by the ``:coq:`` role (use ``:lean3:`` for
+Inline Coq highlighting is provided by the ``:coq:`` role (use ``:lean3:`` or ``:lean4:``for
```
Same here.

insightmind replied:
Removed.

cpitclaudel commented on alectryon/lean4.py:
```diff
@@ -0,0 +1,94 @@
+    LAKE_TMP_FILE_PATH = "lakefile.lean"
+
+    def run_leanInk_document(self, encoded_document):
+        r"""
```
No need for a newline here, nor the `r`

insightmind replied:
Fixed.

cpitclaudel commented on alectryon/lean4.py:
```diff
@@ -0,0 +1,94 @@
+        Run LeanInk with encoded_document file.
+        """
+        with tempfile.TemporaryDirectory(prefix=self.TMP_PREFIX) as temp_directory:
+            inputFile = Path(temp_directory) / os.path.basename(self.fpath.with_suffix(self.LEAN_FILE_EXT))
```
Isn't there a `basename` method on `Path` objects?

insightmind replied:
Not that I know of, I also couldn't find anything about it online.

cpitclaudel replied:
I think it's just `pth.name`?

cpitclaudel commented on alectryon/lean4.py:
```diff
@@ -0,0 +1,94 @@
+    LAKE_ENV_KEY = "--lake"
+    LAKE_TMP_FILE_PATH = "lakefile.lean"
+
+    def run_leanInk_document(self, encoded_document):
```
I've been gradually adding type annotations to Alectryon's codebase; it's your call whether you want to do the same for this file

insightmind replied:
Fixed.

cpitclaudel commented on alectryon/lean4.py:
```diff
@@ -0,0 +1,94 @@
+            return tupleResult
+    
+    def resolve_lake_arg(self):
+        r"""
```
Same here re. `r` and the newline.

insightmind replied:
Fixed.

cpitclaudel commented on alectryon/lean4.py:
```diff
@@ -0,0 +1,94 @@
+
+    def annotate(self, chunks):
+        document = EncodedDocument(chunks, "\n", encoding="utf-8")
+        self.resolve_lake_arg()
```
Is this function needed now that there is a top-level argument this file?

insightmind replied:
Yeah, it is required in order for us to figure out the working directory of LeanInk so all external references are loaded correctly. This behavior is partially handled by the Lean 4 driver.

cpitclaudel commented on alectryon/lean4.py:
```diff
@@ -0,0 +1,94 @@
+        self.resolve_lake_arg()
+        result = self.run_leanInk_document(document)
+
+        if not result:
```
What does this case correspond to?

insightmind replied:
It's a guard statement to prevent an error that would happen when accessing the last element of an empty array.

cpitclaudel commented on alectryon/lean4.py:
```diff
@@ -0,0 +1,94 @@
+        last = result[-1]
+        
+        # Sometimes we require an additonal \n and sometimes not. I wasn't really able to
+        # find out exactly when, but this workaround seems to work for almost all cases.
```
Would it work to ensure that there's a newline at the end of every fragment?

insightmind replied:
I'm not sure if that is enough or correct tbh because we can't be sure if a single newline at the end of a fragment is enough, for example if we have a fragment that has two or more new lines at the end.
However LeanInk should ensure that fragments are generated so that there are no missing characters in between them, so these cases shouldn't be an issue.

I suspect that either the encoding or decoding process of LeanInks input or output causes the issue, or the Lean compiler trims the end of the file in some cases, which would cause the missing new line character.

cpitclaudel commented on recipes/_output/literate_lean4.lean.rst:
```diff
@@ -0,0 +1,63 @@
+        -o literate_lean4.xe.tex
+       # Coq+reST → LaTeX; produces ‘literate_lean4.xe.tex’
+   alectryon literate_lean4.lean --backend rst
+       # Coq+reST → reST;  produces ‘literate_lean4.lean.rst’
```
Could we update the mixed-languages document to show lean4 instead of lean3?

insightmind replied:
Updated the mixed-language document `polyglot.rst` to use Lean 4 instead of Lean 3.

cpitclaudel replied:
Thanks!

cpitclaudel reviewed (approved):
This looks excellent!  I've been delaying this review for much too long because I thought it'd need quite a bit of time, and it took just 30 minutes to go through everything.  Congrats!!

I left a few small comments.  None of them should be time consuming to fix up.

Have you run the code through `make lint` to see if any complaints pop up?

Also, let's update the changelog, and make a release soon after we merge this.

Again, exceptional work! Thanks :)

insightmind commented:
Thank you so much for your thorough review.

I resolved the linter warnings, and adjusted the code based on your review + responded to the comments.

I have also added two new transforms in the `transform.py` file for Lean 4.
This is because of a recent change within LeanInk  to support nested tactics.
Therefore I also regenerated the output files of all Lean 4 recipe examples.

Also let me know if there's anything else that might need some adjustments.

P.S.: Do you want to update the changelog after the merge or should I already update it in this PR?

Kha commented on alectryon/literate.py:
```diff
@@ -862,15 +862,32 @@ def rst2lean3(rst):
+LEAN4 = LangDef(
+    "lean4",
+    LeanParser, # We can use the same parser as Lean 3, because the syntax for comments has not changed between the versions.
+    lit_open=r"/-|", lit_close=r"-/",
```
Lean (3 & 4) uses `/-!` for "module docstrings" that, as it turns out, can be used more than once in a module. After some deliberation with other maintainers, we came to the conclusion that it would be best to reuse this syntax for literate documents instead of introducing a new syntax.

cpitclaudel replied:
Sounds good!  This is specifically for module docstrings, right, not for function docstrings?  The reason I'm asking is that I think it's good to treat literate comments and docstrings differently, since you want to show docstrings e.g. when doing code completion

Kha replied:
Correct, function docstrings is `/--`

cpitclaudel commented on recipes/literate_lean4.lean:
```diff
@@ -44,7 +44,7 @@ By default, these results are folded and are displayed upon hovering or clicking
 #print Iff /- .unfold .no-in -/
 
-/-|
+/-!
```
Didn't you want to change the other instances in this file, as in e.g. above `/-| Other flags`?

insightmind replied:
Yeah I forgot that. I also forgot to adjust the Lean 3 version.
It should now be correct.

cpitclaudel commented on README.rst:
```diff
@@ -483,76 +483,77 @@ See `<recipes/alectryon_custom_driver.py>`__ for a concrete example.
-Alectryon has preliminary support for Lean 3.
-
-Recording Lean's output and generating HTML or LaTeX is supported, from plain ``.lean`` files and from ``.rst`` files using the ``.. lean3::`` directive (as well as Markdown/MyST files using the ``{lean3}`` directive).  Language-agnostic features like caching work.  The literate delimiter is ``/-|``; in other words, you may write:
+Alectryon has support for the Lean 4. LeanInk (`LeanInk <https://github.com/insightmind/LeanInk>`_) is required to use Alectryon with Lean 4 files.
```
```suggestion
Alectryon has support for Lean 4. LeanInk (`LeanInk <https://github.com/insightmind/LeanInk>`_) is required to use Alectryon with Lean 4 files.
```

insightmind replied:
Fixed. I also updated the link to LeanInk because we transferred the ownership of the repo to the Lean organization last week.

cpitclaudel commented on README.rst:
```diff
@@ -483,76 +483,77 @@ See `<recipes/alectryon_custom_driver.py>`__ for a concrete example.
-Recording Lean's output and generating HTML or LaTeX is supported, from plain ``.lean`` files and from ``.rst`` files using the ``.. lean3::`` directive (as well as Markdown/MyST files using the ``{lean3}`` directive).  Language-agnostic features like caching work.  The literate delimiter is ``/-|``; in other words, you may write:
+Alectryon has support for the Lean 4. LeanInk (`LeanInk <https://github.com/insightmind/LeanInk>`_) is required to use Alectryon with Lean 4 files.
+HTML and LaTeX output is supported from plain ``.lean`` source files and from ``.rst`` files.
+The reStructuredText directive for Lean 4 is ``.. lean4::``, for Markdown/MyST files it is ``{lean4}``. The literate delimiter is ``/-|``:
```
```suggestion
The reStructuredText directive for Lean 4 is ``.. lean4::``, for Markdown/MyST files it is ``{lean4}``. The literate delimiter is ``/-!``:
```

insightmind replied:
Fixed.

cpitclaudel commented on README.rst:
```diff
@@ -483,76 +483,77 @@ See `<recipes/alectryon_custom_driver.py>`__ for a concrete example.
+The reStructuredText directive for Lean 4 is ``.. lean4::``, for Markdown/MyST files it is ``{lean4}``. The literate delimiter is ``/-|``:
 
-- Lean 3 code in reStructuredText files, like this:
+- Lean 4 code in reStructuredText files, like this:
```
```suggestion
- Include Lean 4 code in reStructuredText files like this:
```

insightmind replied:
Fixed.

cpitclaudel commented on README.rst:
```diff
@@ -483,76 +483,77 @@ See `<recipes/alectryon_custom_driver.py>`__ for a concrete example.
+        … some Lean 4 code
 
-- reStructuredText prose in Lean3 files, like this:
+- reStructuredText prose in Lean 4 files, like this:
```
```suggestion
- Include reStructuredText prose in Lean 4 files like this:
```

insightmind replied:
Fixed.

cpitclaudel commented on README.rst:
```diff
@@ -483,76 +483,77 @@ See `<recipes/alectryon_custom_driver.py>`__ for a concrete example.
+Alectryon has preliminary support for Lean 3.
 
-- Lean 4 code in reStructuredText files, like this:
+Recording Lean's output and generating HTML or LaTeX is supported, from plain ``.lean`` files and from ``.rst`` files using the ``.. lean3::`` directive (as well as Markdown/MyST files using the ``{lean3}`` directive).  Language-agnostic features like caching work.  The literate delimiter is ``/-|``; in other words, you may write:
```
```suggestion
Recording Lean's output and generating HTML or LaTeX is supported, from plain ``.lean`` files and from ``.rst`` files using the ``.. lean3::`` directive (as well as Markdown/MyST files using the ``{lean3}`` directive).  Language-agnostic features like caching work.  The literate delimiter is ``/-!``; in other words, you may write:
```

insightmind replied:
Fixed.

cpitclaudel reviewed (approved):
OK, a few very minor things, and I think we're ready to merge :)  Thanks again for this excellent work!

cpitclaudel commented:
Hi folks, I finally found the time to go through the PR for a final cleanup, and to merge it! :)  I had to rebase a few things, so I saved a copy of the unrebased branch to `lean4-before-rebase` in this repository.

@insightmind , big congratulations! :)  And @Kha and @insightmind , thanks both a lot for your patience. 

@insightmind , I've sent you an invitation to get you the commit bit, so if you want to make further changes to the lean part you should feel free to either make a PR or push directly to master.  

Cheers!

cpitclaudel commented:
I've installed Lean4 locally to test things, so I'm fairly confident that they work, but I'd love a last go-ahead from either of you before I make a release ^^

Kha commented:
`diff` says output is unchanged on the Lean 4 doc examples, looks good :)
